### ISSUE TITLE
Add new algorithm

### ISSUE THREAD

**Issue by Pikus16** [Pikus16/composer#1]
@droid I want to add a new version of SAM called SAF. Ask me any questions about things that are unclear.

The algorithm is below

```
Input: Training set S; A network fθ with weights θ; Learning rate η; Epochs E; Iterations T per
epoch; SAF starting epoch Estart; SAF coefficients λ; Temperature τ ; SAF hyperparameter E˜;
EMA decay factor β for MESA.

1: for e = 1 to E (where e represents the current epoch) do
  2: for t = 1 to T, Sample a mini-batch B ⊂ S do the following:
     3: Record the outputs: yˆe_i ← fθ(xi), where x_i ∈ B
     4: Load y_iˆ(e−E˜) saved in E˜ epochs ago for each x_i ∈ B
     5: Compute KL divergence between the output at the current epoch ( yˆe_i ) and the previously saved output  y_iˆ(e−E˜) and save that as the trajectory loss 
     6: if e > Estart (Added the trajectory loss after Estart epoch) then
            7. loss = λ * trajectory loss + cross entropy loss
     8. else loss = cross entropy loss
    
   9: Update the weights: θ ← θ − η∇θL
Output: A flat minimum solution ˜θ.
```

**Comment by Pikus16** [comment]
@droid for your questions

- ignore the `EMA decay factor β for MESA`
- use scipy to calculate the KL divergence
- ignore the cross entropy loss term
- `SAF hyperparameter E˜` is how many epochs to look back at when computing the KL divergence. let's just set it to 1

Your plan looks good!
